### PR TITLE
Shell flyout content template

### LINKED ISSUE #6293: [Spec] Shell Flyout Content Template
# Flyout Content Template

Allow developers to provide a template for the entire flyout.

* Should be able to describe the whole layout
* Developer responsible for header and footer. Any provided templates would not be applied.

```
<Shell.FlyoutTemplate>
    <DataTemplate>
        <Grid>
            <ContentView>
                <Label Text="My Header"/>
            </ContentView>
            
            <ListView ItemsSource="{Binding Items}">
                <ListView.ItemTemplate>
                    <DataTemplate>
                        <Grid>
                            <Image Source="{Binding Icon}"/>
                            <Label Text="{Binding Title}"/>
                        </Grid>
                    </DataTemplate>
                </ListView.ItemTemplate>
            </ListView>

            <ContentView>
                <Label Text="My Footer"/>
            </ContentView>
        </Grid>
    </DataTemplate>
</Shell.FlyoutTemplate>
```

# API

Shell.FlyoutTemplate

# Difficulty : Awesome

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

Only works on ios/android

This adds the ability to specify your own Content for the Flyout. If you don't want the list of FlyoutItems you can now specify Content or a ContentTemplate and it will render instead of the Flyout Items.

If you specify a Header or a Footer those will also renderer. If the content you specify is scrollable Shell will do its best to honor the Scroll Behavior for the header you've specified

### Issues Resolved ### 

- fixes #6293

### API Changes ###

Added:
 - View Shell.FlyoutContent //BP
 - DataTemplate Shell.FlyoutContentTemplate //BP
 - IEnumerable Shell.FlyoutItems //Readonly list of Flyout Items. The underlying list implements INCC via ReadonlyObservableCollection

### Platforms Affected ### 
- Core/XAML (all platforms)

### Testing Procedure ###
- Tests Included

### XAML Example

```XAML
<Shell x:Name=shell>
     <Shell.FlyoutContent>
           <CollectionView 
                    BindingContext="{x:Reference Name=shell}"  
                    IsGrouped=true 
                    ItemSource={Binding FlyoutItems}>
                       <CollectionView.ItemTemplate>
                              <DataTemplate>
                                   <Label Text="{Binding Title}">
                              </DataTemplate>
                       </CollectionView.ItemTemplate>
             </CollectionView>
     </Shell.FlyoutContent>
</Shell>
```

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentManager.cs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+				if (ScrollView != null && !(ScrollView is UITableView))
+					ScrollView.Scrolled -= ScrollViewScrolled;
+
+				_contentView = value;
+
+				if (ScrollView != null && !(ScrollView is UITableView))
+					ScrollView.Scrolled += ScrollViewScrolled;
+
+				if (ScrollView != null && Forms.IsiOS11OrNewer)
+					ScrollView.ContentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentBehavior.Never;
```
Comment (pictos) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentManager.cs:55]:
```suggestion
                                if (ScrollView == null)
                                      return;
				
				if (!(ScrollView is UITableView))
					ScrollView.Scrolled += ScrollViewScrolled;

				if (Forms.IsiOS11OrNewer)
					ScrollView.ContentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentBehavior.Never;
```

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs [hunk 2]
```diff
@@ -14,6 +14,9 @@ public class ShellFlyoutContentRenderer : UIViewController, IShellFlyoutContentR
 		UIView _footerView;
 		View _footer;
 		ShellTableViewController _tableViewController;
+		ShellFlyoutContentManager _shellFlyoutContentManager;
+		private View _flyoutContent;
+		private UIView _flyoutContentView;
```
Comment (pictos) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs:19]:
```suggestion
		View _flyoutContent;
		UIView _flyoutContentView;
```

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs [hunk 3]
```diff
@@ -273,6 +283,54 @@ public override void ViewDidLoad()
...
+			}
+
+			if (_flyoutContent == view)
+				return;
+
+			_tableViewController.View.RemoveFromSuperview();
+
+			_flyoutContent = view;
+
+			if (_flyoutContent != null)
```
Comment (pictos) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs:322]: we don't need to check for null here, do we? we set the `_flyoutContent ` to the `view` value. If `view` is null we return the code at line 312, and we don't get this line.

**Review by pictos — COMMENTED** [review]
I didn't run the code yet, but I did a quick overview of the code and have these suggestions.

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentManager.cs (hunk 1 above)
Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentManager.cs:55]: The content might not be a ScrollView type

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs (hunk 2 above)
Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: deleted this code

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs (hunk 3 above)
Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: Removed this code

**Comment by PureWeen** [comment]
https://dev.azure.com/xamarin/public/_releaseProgress?_a=release-pipeline-progress&releaseId=2692

**Review by rmarinho — APPROVED** [review]

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs [hunk 4]
```diff
@@ -251,10 +258,10 @@ public override void ViewDidLoad()
 		{
 			base.ViewDidLoad();
 
-			View.AddSubview(_tableViewController.View);
 
 			UpdateFlyoutHeader();
 			UpdateFlyoutFooter();
+		//	UpdateFlyoutContent();
```
Comment (hartez) [inline Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs:264]: Is this intentional?

**Review by hartez — CHANGES_REQUESTED** [review]
Wondering about that one Update method that's commented out. Other than that, looks fine (though I wouldn't mind seeing some of these longer methods broken up).

File: Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs (hunk 4 above)
Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Renderers/ShellFlyoutContentRenderer.cs]: Removed

**Review by pictos — APPROVED** [review]
Everything works on my tests.

**Review by hartez — APPROVED** [review]

**Pull request merged**